src/main/java/com/gregtechceu/gtceu/utils/EntityDamageUtil.java

```diff
@@ -0,0 +1,113 @@
+package com.gregtechceu.gtceu.utils;
+
+import com.gregtechceu.gtceu.common.data.GTDamageTypes;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.effect.MobEffectInstance;
+import net.minecraft.world.effect.MobEffects;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.animal.PolarBear;
+import net.minecraft.world.entity.animal.SnowGolem;
+import net.minecraft.world.entity.boss.wither.WitherBoss;
+import net.minecraft.world.entity.monster.*;
+import net.minecraft.world.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public class EntityDamageUtil {
+
+    private static final int FROST_WALKER_ID = 9;
+
+    /**
+     * @param entity      the entity to damage
+     * @param temperature the temperature of the fluid in the pipe
+     * @param multiplier  the multiplier on the damage taken
+     * @param maximum     the maximum damage to apply to the entity, use -1 for no maximum
+     */
+    public static void applyTemperatureDamage(@NotNull LivingEntity entity, int temperature, float multiplier,
+                                              int maximum) {
+        if (temperature > 320) {
+            int damage = (int) ((multiplier * (temperature - 300)) / 50.0F);
+            if (maximum > 0) {
+                damage = Math.min(maximum, damage);
+            }
+            applyHeatDamage(entity, damage);
+        } else if (temperature < 260) {
+            int damage = (int) ((multiplier * (273 - temperature)) / 25.0F);
+            if (maximum > 0) {
+                damage = Math.min(maximum, damage);
+            }
+            applyFrostDamage(entity, damage);
+        }
+    }
+
+    /**
+     * @param entity the entity to damage
+     * @param damage the damage to apply
+     */
+    public static void applyHeatDamage(@NotNull LivingEntity entity, int damage) {
+        // do not attempt to damage by 0
+        if (damage <= 0) return;
+        if (!entity.isAlive()) return;
+        // fire/lava mobs cannot be burned
+        if (entity instanceof Blaze || entity instanceof MagmaCube ||
+                entity instanceof WitherSkeleton || entity instanceof WitherBoss)
```

Comment:
Would probably be better for compatibility to make this a tag (if entities can have tags - I'm not sure about that)